File: analysis/deprecated_analise_modelos_cesarea_corrigido.R
# =============================================================================
# DEPRECATED — 2026-05-23
# Substituído por: analysis/04_modelo4_regressao.R
# Razão: usa BD_completo_corrigido_13-05-2026.xls e filtro ad-hoc (não §3.3).
# O script novo usa a fonte canônica (06-04-2026.xlsx) e filtro §3.3 (N=6.650).
# =============================================================================
# =============================================================================
# ANÁLISE CORRIGIDA: Modelos de Regressão Logística para Cesárea
# Correção: Uso de 'dheg' (variável correta) no lugar de 'dheg_hipertensao_obst'
# Baseado nos 4 modelos sugeridos pela estatística (arquivo: Resultados_finais.ods)
# =============================================================================
# Autora: Letícia Schimidt Arruda (Dissertação de Mestrado)
# Data: 2026-05-22
# Linguagem: R
# =============================================================================

library(tidyverse)
library(readxl)
library(here)

# -----------------------------------------------------------------------------
# 1. LEITURA E FILTRO DOS DADOS
# -----------------------------------------------------------------------------
dados <- read_excel(
  here("data", "raw", "BD_completo_corrigido_13-05-2026.xls"),
  sheet = "BD_leticia_08-05"
)

# Filtros idênticos ao script original (2025_03_29_analisesGestantes.R)
dadosFilt <- dados %>%
  filter(
    !is.na(tipo_parto),
    !is.na(idade),
    idade > 10,
    idade < 35,
    !(origem == "Adolescentes" & idade == 20)
  )

cat("=== Tamanho da amostra após filtros:", nrow(dadosFilt), "registros ===\n\n")

# -----------------------------------------------------------------------------
# 2. PREPARAÇÃO DAS VARIÁVEIS
# Código 999 = valor ausente em todas as variáveis categóricas
# -----------------------------------------------------------------------------
prep_var <- function(x) {
  x[x == 999] <- NA
  x
}

dadosFilt <- dadosFilt %>%
  mutate(
    # Desfecho: parto cesárea (0 = não, 1 = sim)
    cesarea       = prep_var(parto_cesarea),

    # Variável principal corrigida: dheg (0 = não, 1 = sim)
    # SPSS usou incorretamente: dheg_hipertensao_obst
    dheg          = prep_var(dheg),

    # Variável INCORRETA do SPSS (mantida para comparação)
    dheg_spss     = prep_var(dheg_hipertensao_obst),

    # Faixa etária: 1 = Adolescente (ref), 2 = Adulta
    faixa_etaria  = factor(prep_var(faixa_etaria_2cat),
                           levels = c(1, 2),
                           labels = c("Adolescente", "Adulta")),

    # Apresentação fetal: 1 = Cefálica (ref), 2 = Pélvica/anômala
    apres_feto_bin = case_when(
      prep_var(apres_feto) == 1 ~ 0,  # Cefálica
      prep_var(apres_feto) == 2 ~ 1,  # Não-cefálica
      TRUE ~ NA_real_
    ),

    # Diabetes Gestacional Obstétrico: 0 = não, 1 = sim
    dmg_obst      = prep_var(dmg_obst),

    # Grupo de Robson Reduzido (fator com ref = grupo 1)
    robson        = factor(prep_var(Robson_reduzido))
  ) %>%
  mutate(
    robson = relevel(robson, ref = "1")
  )

# -----------------------------------------------------------------------------
# 3. FUNÇÕES AUXILIARES
# -----------------------------------------------------------------------------

# Extrai tabela de OR, IC95% e p-valor de um glm
extract_or_table <- function(model, nome_modelo) {
  coef_df   <- summary(model)$coefficients
  conf_int  <- suppressMessages(confint(model))

  result <- data.frame(
    Modelo    = nome_modelo,
    Variavel  = rownames(coef_df),
    B         = round(coef_df[, "Estimate"], 3),
    p_valor   = round(coef_df[, "Pr(>|z|)"], 4),
    OR        = round(exp(coef_df[, "Estimate"]), 2),
    IC95_inf  = round(exp(conf_int[, 1]), 2),
    IC95_sup  = round(exp(conf_int[, 2]), 2),
    stringsAsFactors = FALSE
  )
  return(result)
}

# Teste de Hosmer-Lemeshow manual
hosmer_lemeshow <- function(model, g = 10) {
  obs  <- model$y
  pred <- fitted(model)
  df_hl <- data.frame(obs, pred) %>%
    arrange(pred) %>%
    mutate(grupo = ntile(pred, g))

  hl_tab <- df_hl %>%
    group_by(grupo) %>%
    summarise(
      obs_sim  = sum(obs),
      obs_nao  = n() - sum(obs),
      esp_sim  = sum(pred),
      esp_nao  = n() - sum(pred),
      .groups = "drop"
    )

  chi2 <- sum(
    (hl_tab$obs_sim  - hl_tab$esp_sim)^2  / hl_tab$esp_sim  +
    (hl_tab$obs_nao  - hl_tab$esp_nao)^2  / hl_tab$esp_nao
  )
  df_hl_stat <- g - 2
  p_hl <- pchisq(chi2, df = df_hl_stat, lower.tail = FALSE)
  cat(sprintf("  Hosmer-Lemeshow: χ²=%.3f, df=%d, p=%.3f\n", chi2, df_hl_stat, p_hl))
  invisible(list(chi2 = chi2, df = df_hl_stat, p = p_hl))
}

# Nagelkerke R²
nagelkerke_r2 <- function(model) {
  n       <- length(model$y)
  ll_null <- logLik(update(model, . ~ 1))
  ll_full <- logLik(model)
  r2_cs   <- 1 - exp((2/n) * (as.numeric(ll_null) - as.numeric(ll_full)))
  r2_max  <- 1 - exp((2/n) * as.numeric(ll_null))
  round(r2_cs / r2_max, 4)
}

# -----------------------------------------------------------------------------
# 4. MODELOS CORRIGIDOS (com dheg)
# -----------------------------------------------------------------------------
cat("============================================================\n")
cat("  MODELOS CORRIGIDOS — usando variável 'dheg' (correta)\n")
cat("============================================================\n\n")

# --- Sugestão 1: Clínico Clássico ---
cat("--- Sugestão 1: Faixa Etária + dheg + Apresentação Fetal ---\n")
df_m1 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg, apres_feto_bin) %>%
  drop_na()
cat("  N =", nrow(df_m1), "\n")
m1 <- glm(cesarea ~ faixa_etaria + dheg + apres_feto_bin,
          data = df_m1, family = binomial)
hosmer_lemeshow(m1)
cat("  Nagelkerke R² =", nagelkerke_r2(m1), "\n")
print(extract_or_table(m1, "Sugestão 1 (corrigida)"))
cat("\n")

# --- Sugestão 2: Clínico + DMG ---
cat("--- Sugestão 2: Faixa Etária + dheg + Apresentação Fetal + DMG ---\n")
df_m2 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg, apres_feto_bin, dmg_obst) %>%
  drop_na()
cat("  N =", nrow(df_m2), "\n")
m2 <- glm(cesarea ~ faixa_etaria + dheg + apres_feto_bin + dmg_obst,
          data = df_m2, family = binomial)
hosmer_lemeshow(m2)
cat("  Nagelkerke R² =", nagelkerke_r2(m2), "\n")
print(extract_or_table(m2, "Sugestão 2 (corrigida)"))
cat("\n")

# --- Sugestão 3: Robson + Apresentação Fetal (sobrecarga) ---
cat("--- Sugestão 3: Faixa Etária + dheg + Apresentação Fetal + Robson ---\n")
df_m3 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg, apres_feto_bin, robson) %>%
  drop_na()
cat("  N =", nrow(df_m3), "\n")
m3 <- glm(cesarea ~ faixa_etaria + dheg + apres_feto_bin + robson,
          data = df_m3, family = binomial)
hosmer_lemeshow(m3)
cat("  Nagelkerke R² =", nagelkerke_r2(m3), "\n")
print(extract_or_table(m3, "Sugestão 3 (corrigida)"))
cat("\n")

# --- Sugestão 4: Robson Consolidado (modelo recomendado) ---
cat("--- Sugestão 4: Faixa Etária + dheg + Robson (MODELO RECOMENDADO) ---\n")
df_m4 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg, robson) %>%
  drop_na()
cat("  N =", nrow(df_m4), "\n")
m4 <- glm(cesarea ~ faixa_etaria + dheg + robson,
          data = df_m4, family = binomial)
hosmer_lemeshow(m4)
cat("  Nagelkerke R² =", nagelkerke_r2(m4), "\n")
print(extract_or_table(m4, "Sugestão 4 (corrigida)"))
cat("\n")

# -----------------------------------------------------------------------------
# 5. MODELOS COM dheg_hipertensao_obst (SPSS — para comparação)
# -----------------------------------------------------------------------------
cat("============================================================\n")
cat("  MODELOS SPSS — usando 'dheg_hipertensao_obst' (INCORRETO)\n")
cat("============================================================\n\n")

cat("--- Sugestão 1 (SPSS): Faixa Etária + dheg_hipertensao_obst + Apresentação Fetal ---\n")
df_s1 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg_spss, apres_feto_bin) %>%
  drop_na()
cat("  N =", nrow(df_s1), "\n")
s1 <- glm(cesarea ~ faixa_etaria + dheg_spss + apres_feto_bin,
          data = df_s1, family = binomial)
hosmer_lemeshow(s1)
cat("  Nagelkerke R² =", nagelkerke_r2(s1), "\n")
print(extract_or_table(s1, "Sugestão 1 (SPSS/incorreta)"))
cat("\n")

cat("--- Sugestão 4 (SPSS): Faixa Etária + dheg_hipertensao_obst + Robson ---\n")
df_s4 <- dadosFilt %>%
  select(cesarea, faixa_etaria, dheg_spss, robson) %>%
  drop_na()
cat("  N =", nrow(df_s4), "\n")
s4 <- glm(cesarea ~ faixa_etaria + dheg_spss + robson,
          data = df_s4, family = binomial)
hosmer_lemeshow(s4)
cat("  Nagelkerke R² =", nagelkerke_r2(s4), "\n")
print(extract_or_table(s4, "Sugestão 4 (SPSS/incorreta)"))
cat("\n")

# -----------------------------------------------------------------------------
# 6. SALVAR RESULTADOS
# -----------------------------------------------------------------------------
all_results <- bind_rows(
  extract_or_table(m1, "Sug1_corrigida"),
  extract_or_table(m2, "Sug2_corrigida"),
  extract_or_table(m3, "Sug3_corrigida"),
  extract_or_table(m4, "Sug4_corrigida"),
  extract_or_table(s1, "Sug1_SPSS_incorreta"),
  extract_or_table(s4, "Sug4_SPSS_incorreta")
)

output_path <- here("results", "tabelas_dissertacao", "modelos_cesarea_corrigidos_R.csv")
write_csv(all_results, output_path)
cat("Resultados salvos em:", output_path, "\n")

File: analysis/deprecated_analise_modelos_cesarea_corrigido.py
"""
=============================================================================
DEPRECATED — 2026-05-23
Substituído por: analysis/04_modelo4_regressao.R
Razão: usa BD_completo_corrigido_13-05-2026.xls e filtro ad-hoc (não §3.3).
=============================================================================
=============================================================================
ANÁLISE CORRIGIDA: Modelos de Regressão Logística para Cesárea
Correção: Uso de 'dheg' (variável correta) no lugar de 'dheg_hipertensao_obst'
Baseado nos 4 modelos sugeridos pela estatística (arquivo: Resultados_finais.ods)
=============================================================================
Autora: Letícia Schimidt Arruda (Dissertação de Mestrado)
Data: 2026-05-22
Linguagem: Python 3
Dependências: pandas, numpy, statsmodels, scipy, openpyxl/xlrd
=============================================================================
"""

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from scipy import stats
from pathlib import Path

# -----------------------------------------------------------------------------
# 1. CAMINHOS
# -----------------------------------------------------------------------------
BASE_DIR = Path(__file__).resolve().parent.parent
DATA_FILE = BASE_DIR / "data" / "raw" / "BD_completo_corrigido_13-05-2026.xls"
OUTPUT_DIR = BASE_DIR / "results" / "tabelas_dissertacao"
OUTPUT_FILE = OUTPUT_DIR / "modelos_cesarea_corrigados_Python.csv"

# -----------------------------------------------------------------------------
# 2. LEITURA E FILTRO DOS DADOS
# -----------------------------------------------------------------------------
dados = pd.read_excel(DATA_FILE, sheet_name="BD_leticia_08-05")

mask = (
    dados["tipo_parto"].notna() &
    dados["idade"].notna() &
    (dados["idade"] > 10) &
    (dados["idade"] < 35) &
    ~((dados["origem"] == "Adolescentes") & (dados["idade"] == 20))
)
df = dados[mask].copy()
print(f"=== Tamanho da amostra após filtros: {len(df)} registros ===\n")

# -----------------------------------------------------------------------------
# 3. PREPARAÇÃO DAS VARIÁVEIS (999 → NaN)
# -----------------------------------------------------------------------------
for col in ["parto_cesarea", "dheg", "dheg_hipertensao_obst",
            "faixa_etaria_2cat", "apres_feto", "dmg_obst", "Robson_reduzido"]:
    df[col] = df[col].replace(999, np.nan)

# Desfecho
df["cesarea"] = df["parto_cesarea"]

# Variável corrigida: dheg
df["dheg_corr"] = df["dheg"]

# Variável SPSS (incorreta)
df["dheg_spss"] = df["dheg_hipertensao_obst"]

# Faixa etária: 0 = Adolescente (ref), 1 = Adulta
df["adulta"] = (df["faixa_etaria_2cat"] == 2).astype(float)
df.loc[df["faixa_etaria_2cat"].isna(), "adulta"] = np.nan

# Apresentação fetal anômala: 0 = Cefálica (ref), 1 = Não-cefálica
df["apres_anomala"] = np.where(df["apres_feto"] == 2, 1.0,
                      np.where(df["apres_feto"] == 1, 0.0, np.nan))

# Robson Reduzido como categórico (grupo 1 = referência)
df["robson"] = pd.Categorical(df["Robson_reduzido"], ordered=False)

# -----------------------------------------------------------------------------
# 4. FUNÇÕES AUXILIARES
# -----------------------------------------------------------------------------

def hosmer_lemeshow(y_obs, y_pred, g=10):
    """Calcula estatística de Hosmer-Lemeshow."""
    df_hl = pd.DataFrame({"obs": y_obs, "pred": y_pred}).dropna()
    df_hl["decil"] = pd.qcut(df_hl["pred"], q=g, labels=False, duplicates="drop")
    grouped = df_hl.groupby("decil").agg(
        obs_sim=("obs", "sum"),
        n=("obs", "count"),
        esp_sim=("pred", "sum")
    ).reset_index()
    grouped["obs_nao"] = grouped["n"] - grouped["obs_sim"]
    grouped["esp_nao"] = grouped["n"] - grouped["esp_sim"]
    # Evitar divisão por zero
    grouped = grouped[(grouped["esp_sim"] > 0) & (grouped["esp_nao"] > 0)]
    chi2 = (
        ((grouped["obs_sim"] - grouped["esp_sim"]) ** 2 / grouped["esp_sim"]) +
        ((grouped["obs_nao"] - grouped["esp_nao"]) ** 2 / grouped["esp_nao"])
    ).sum()
    df_stat = len(grouped) - 2
    p_val = 1 - stats.chi2.cdf(chi2, df=df_stat)
    return chi2, df_stat, p_val


def nagelkerke_r2(result):
    """Calcula Nagelkerke R² a partir do resultado do statsmodels."""
    n = result.nobs
    ll_null = result.llnull
    ll_full = result.llf
    r2_cs = 1 - np.exp((2 / n) * (ll_null - ll_full))
    r2_max = 1 - np.exp((2 / n) * ll_null)
    return round(r2_cs / r2_max, 4)


def extract_results(result, nome_modelo):
    """Extrai OR, IC95%, p-valor de um resultado de regressão logística."""
    coef = result.params
    pval = result.pvalues
    conf = result.conf_int()
    rows = []
    for var in coef.index:
        rows.append({
            "Modelo":   nome_modelo,
            "Variavel": var,
            "B":        round(coef[var], 3),
            "p_valor":  round(pval[var], 4),
            "OR":       round(np.exp(coef[var]), 2),
            "IC95_inf": round(np.exp(conf.loc[var, 0]), 2),
            "IC95_sup": round(np.exp(conf.loc[var, 1]), 2),
        })
    return pd.DataFrame(rows)


def fit_logit(formula, data, nome_modelo, drop_cols=None):
    """Ajusta regressão logística, imprime resultados e retorna DataFrame."""
    if drop_cols:
        df_model = data.dropna(subset=drop_cols)
    else:
        df_model = data
    result = smf.logit(formula, data=df_model).fit(disp=False)
    n = int(result.nobs)
    print(f"--- {nome_modelo} --- N={n}")
    chi2_hl, df_hl, p_hl = hosmer_lemeshow(
        result.model.endog,
        result.predict()
    )
    print(f"  Hosmer-Lemeshow: χ²={chi2_hl:.3f}, df={df_hl}, p={p_hl:.3f}")
    print(f"  Nagelkerke R² = {nagelkerke_r2(result)}")
    tbl = extract_results(result, nome_modelo)
    print(tbl.to_string(index=False))
    print()
    return tbl, result


# -----------------------------------------------------------------------------
# 5. MODELOS CORRIGIDOS (com dheg_corr)
# -----------------------------------------------------------------------------
print("=" * 60)
print("  MODELOS CORRIGIDOS — usando 'dheg' (correto)")
print("=" * 60 + "\n")

all_tables = []

# Sugestão 1
t1, r1 = fit_logit(
    "cesarea ~ adulta + dheg_corr + apres_anomala",
    df, "Sug1_corrigida",
    drop_cols=["cesarea", "adulta", "dheg_corr", "apres_anomala"]
)
all_tables.append(t1)

# Sugestão 2
t2, r2 = fit_logit(
    "cesarea ~ adulta + dheg_corr + apres_anomala + dmg_obst",
    df, "Sug2_corrigida",
    drop_cols=["cesarea", "adulta", "dheg_corr", "apres_anomala", "dmg_obst"]
)
all_tables.append(t2)

# Sugestão 3
t3, r3 = fit_logit(
    "cesarea ~ adulta + dheg_corr + apres_anomala + C(robson, Treatment(reference=1.0))",
    df, "Sug3_corrigida",
    drop_cols=["cesarea", "adulta", "dheg_corr", "apres_anomala", "robson"]
)
all_tables.append(t3)

# Sugestão 4
t4, r4 = fit_logit(
    "cesarea ~ adulta + dheg_corr + C(robson, Treatment(reference=1.0))",
    df, "Sug4_corrigida",
    drop_cols=["cesarea", "adulta", "dheg_corr", "robson"]
)
all_tables.append(t4)

# -----------------------------------------------------------------------------
# 6. MODELOS SPSS (com dheg_hipertensao_obst — para comparação)
# -----------------------------------------------------------------------------
print("=" * 60)
print("  MODELOS SPSS — usando 'dheg_hipertensao_obst' (INCORRETO)")
print("=" * 60 + "\n")

ts1, rs1 = fit_logit(
    "cesarea ~ adulta + dheg_spss + apres_anomala",
    df, "Sug1_SPSS_incorreta",
    drop_cols=["cesarea", "adulta", "dheg_spss", "apres_anomala"]
)
all_tables.append(ts1)

ts4, rs4 = fit_logit(
    "cesarea ~ adulta + dheg_spss + C(robson, Treatment(reference=1.0))",
    df, "Sug4_SPSS_incorreta",
    drop_cols=["cesarea", "adulta", "dheg_spss", "robson"]
)
all_tables.append(ts4)

# -----------------------------------------------------------------------------
# 7. SALVAR RESULTADOS
# -----------------------------------------------------------------------------
result_df = pd.concat(all_tables, ignore_index=True)
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
result_df.to_csv(OUTPUT_FILE, index=False)
print(f"Resultados salvos em: {OUTPUT_FILE}")
